src/close.rs

```diff
@@ -0,0 +1,99 @@
+use crate::{transaction::CloseTransaction, Channel};
+use bitcoin::{Address, Transaction};
+use ecdsa_fun::Signature;
+
+pub struct State0(Channel);
+pub struct State1 {
+    channel: Channel,
+    final_address_self: Address,
+    final_address_other: Address,
+}
+pub struct FinalState(Transaction);
+
+impl FinalState {
+    pub fn into_transaction(self) -> Transaction {
+        self.0
+    }
+}
+
+impl From<Channel> for State0 {
+    fn from(channel: Channel) -> Self {
+        Self(channel)
+    }
+}
+
+pub struct Message0 {
+    final_address: Address,
+}
+
+#[derive(Debug)]
+pub struct Message1 {
+    close_transaction: CloseTransaction,
+    sig_close_transaction: Signature,
+}
+
+impl State0 {
+    pub fn compose(&self, final_address: Address) -> Message0 {
+        Message0 { final_address }
+    }
+
+    pub fn interpret(
+        self,
+        final_address_self: Address,
+        Message0 {
+            final_address: final_address_other,
+        }: Message0,
+    ) -> State1 {
+        State1 {
+            channel: self.0,
+            final_address_self,
+            final_address_other,
+        }
+    }
+}
+
+impl State1 {
+    pub fn compose(&self) -> anyhow::Result<Message1> {
+        let (amount_a, ownership_a) = self.channel.current_state.signed_TX_s.outputs().a;
+        let (amount_b, ownership_b) = self.channel.current_state.signed_TX_s.outputs().b;
+
+        let (output_a, output_b) = if ownership_a == self.channel.x_self.public() {
+            (
+                (amount_a, self.final_address_self.clone()),
+                (amount_b, self.final_address_other.clone()),
+            )
+        } else if ownership_b == self.channel.x_self.public() {
+            (
+                (amount_a, self.final_address_other.clone()),
+                (amount_b, self.final_address_self.clone()),
+            )
+        } else {
+            anyhow::bail!("No valid output found")
+        };
```

Comment:
:wrench: The `CloseTransaction` could be pre-computed and saved as part of receiving the previous message. That way you avoid having to send the `CloseTransaction` between the parties - they can just use their own.

Reply:
Would be possible. 
Do you think it has benefits to send over the `CloseTransaction` so that both parties can ensure that they generated the same?

Reply:
I don't see and advantage over verifying the signature they send over: if the signature verifies, they signed the correct transaction (the same one as us) with respect to the correct `OwnershipPublicKey`, so they _must_ have generated the same transaction.

Reply:
True that :)